tests/std/tests/GH_003022_substr_allocator/test.cpp

```diff
@@ -0,0 +1,204 @@
+        {STATICALLY_WIDEN(CharT, "Short substring"), 9, 3},
+        {STATICALLY_WIDEN(CharT, "Mid substring"), 7, 5},
+        {STATICALLY_WIDEN(CharT, "Longer string"), 3, 9},
+        {STATICALLY_WIDEN(CharT, "Maximal SSO len"), 0, 15},
+        {STATICALLY_WIDEN(CharT, "Take an empty substring from a long string"), 35, 0},
+        {STATICALLY_WIDEN(CharT, "Take a short substring from a long string"), 21, 2},
+        {STATICALLY_WIDEN(CharT, "Take some substring from a long string"), 15, 6},
+        {STATICALLY_WIDEN(CharT, "Take a longer substring from a long string"), 20, 13},
+        {STATICALLY_WIDEN(CharT, "Take a quiet longer substring..."), 10, 20},
+        {STATICALLY_WIDEN(CharT, "Take the whole long string as a substring"), 0, 41},
+    };
+
+    using string_type = basic_string<CharT, char_traits<CharT>, payloaded_allocator<CharT>>;
+
+    auto myator = payloaded_allocator<CharT>{42};
+
+    for (const auto& test_case : substring_test_cases) {
+        auto full_str = string_type{test_case.ntcts, myator};
+
+        // Test const lvalue overloads
+        TEST_ASSERT(string_type{full_str, test_case.offset, test_case.count}.get_allocator().get_payload() == 0);
+        TEST_ASSERT(string_type{full_str, test_case.offset}.get_allocator().get_payload() == 0);
+
+        TEST_ASSERT(full_str.substr(test_case.offset, test_case.count).get_allocator().get_payload() == 0);
+
+        TEST_ASSERT(
+            string_type{full_str, test_case.offset, test_case.count, myator}.get_allocator().get_payload() == 42);
+        TEST_ASSERT(string_type{full_str, test_case.offset, myator}.get_allocator().get_payload() == 42);
+
+        // Test non-const rvalue overloads
+        TEST_ASSERT(
+            string_type{string_type{full_str}, test_case.offset, test_case.count}.get_allocator().get_payload() == 0);
+        TEST_ASSERT(string_type{string_type{full_str}, test_case.offset}.get_allocator().get_payload() == 0);
+
+        TEST_ASSERT(string_type{full_str}.substr(test_case.offset, test_case.count).get_allocator().get_payload() == 0);
+
+        TEST_ASSERT(
+            string_type{string_type{full_str}, test_case.offset, test_case.count, myator}.get_allocator().get_payload()
+            == 42);
+        TEST_ASSERT(string_type{string_type{full_str}, test_case.offset, myator}.get_allocator().get_payload() == 42);
+    }
+
+    // Also test well-formedness change before/after P2438R2
+    struct ambiguator {
+        string_type str;
+
+        operator const string_type&() {
+            return str;
+        }
+
+        operator string_type&&() {
+            return move(str);
+        }
+    };
+
+#if _HAS_CXX23 && !defined(__EDG__) // TRANSITION, EDG doesn't perform overload resolution correctly
+    static_assert(!is_constructible_v<string_type, ambiguator&, size_t, size_t>, "Ambiguous");
+#else
+    static_assert(is_constructible_v<string_type, ambiguator&, size_t, size_t>, "Not ambiguous");

```

Comment:
Suggestion: These messages are slightly confusing as to whether they're describing what *should* happen, or what actually happened if they have failed.